src/AutorestSwift/View Models/OperationViewModel.swift

```diff
@@ -63,27 +63,18 @@ struct OperationViewModel {
         var requiredHeaders = [KeyValueViewModel]()
         var optionalQueryParams = [KeyValueViewModel]()
         var optionalHeaders = [KeyValueViewModel]()
-        var uriParams = [KeyValueViewModel]()
+        var pathParams = [KeyValueViewModel]()
         var pipelineContext = [KeyValueViewModel]()
 
-        for param in operation.parameters ?? [] {
-            guard let httpParam = param.protocol.http as? HttpParameter else { continue }
-
-            let viewModel = KeyValueViewModel(from: param, with: operation)
-
-            switch httpParam.in {
-            case .query:
-                viewModel.optional ? optionalQueryParams.append(viewModel) : requiredQueryParams.append(viewModel)
-            case .header:
-                viewModel.optional ? optionalHeaders.append(viewModel) : requiredHeaders
-                    .append(viewModel)
-            case .path,
-                 .uri:
-                uriParams.append(viewModel)
-            default:
-                continue
-            }
-        }
+        buildQueryParamsHeaders(
+            parameters: operation.parameters,
+            with: operation,
+            requiredQueryParams: &requiredQueryParams,
+            requiredHeaders: &requiredHeaders,
+            optionalQueryParams: &optionalQueryParams,
+            optionalHeaders: &optionalHeaders,
+            pathParams: &pathParams
+        )
```

Comment:
Seems weird to have one parameter that is `with`. Might make more sense to just use `operation` like the others. The use of `inout` syntax is certainly atypical but is, I'm assuming, a result of using structs. A cleaner implementation might be to have a single utility struct that holds all the useful info, and then you can pass in the before and return the after:
```swift
struct OperationParameters {
   let header: Params
   let query: Params
   let path: Params
}

struct Params {
  let required: [String:String]
  let optional: [String:String]
}

let finalParams = buildQueryParamsHeaders(parameters:...,operation:..., params: params)
```

Then you could just dot into what you want `for param in params.header.required { ... }`

Reply:
Thanks for the suggestion. Changed per suggestion.